drivers: intc_plic: refactor to prepare for multi-instance support

ycsin opened the pull request:
- Introduce a config struct to store per-instance device config during init and connect the IRQ based on the devicetree instead of hardcoded values.
- The PLIC registers are now defined  directly in the driver, instead of the devicetree as named registers
  - The `trigger type` register isn't defined in the RISCV PLIC specs, but is defined in Telink's & Andes's datasheet
- Fix `telink_b91` compilation error - #62542 @yurvyn
  - [telink datasheet](http://wiki.telink-semi.cn/doc/ds/DS-TLSR9518-E_Datasheet%20for%20Telink%20Multi-Standard%20Wireless%20SoC%20TLSR9518.pdf#page=107&zoom=100,70,76)

Note: The `get_plic_dev_from_irq` is still a placeholder for now and always return the first instance.

This PR serves as a base and will make the multi-instance implementation simpler to address #62441 later.

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -21,24 +21,39 @@
+ *
+ * @return PLIC device of that IRQ
+ */
+static const struct device *get_plic_dev_from_irq(uint32_t irq)
```
```suggestion
static inline const struct device *get_plic_dev_from_irq(uint32_t irq)
```

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -52,8 +67,11 @@ static int save_irq;
+
 	if (IS_ENABLED(CONFIG_PLIC_SUPPORTS_EDGE_IRQ)) {
-		volatile uint32_t *trig = (volatile uint32_t *)PLIC_EDGE_TRIG_TYPE;
+		volatile uint32_t *trig = (volatile uint32_t *)config->trig;
```
here and everywhere a space after casting:
```suggestion
		volatile uint32_t *trig = (volatile uint32_t *) config->trig;
```
pet peeve of mine

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -208,43 +233,64 @@ static void plic_irq_handler(const void *arg)
 
 	/* Set priority of each interrupt line to 0 initially */
-	for (i = 0; i < PLIC_IRQS; i++) {
+	for (int i = 0; i < config->num_irqs; i++) {
```
```suggestion
	for (size_t i = 0; i < config->num_irqs; i++) {
```

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -208,43 +233,64 @@ static void plic_irq_handler(const void *arg)
 
 	/* Ensure that all interrupts are disabled initially */
-	for (i = 0; i < PLIC_EN_SIZE; i++) {
+	for (int i = 0; i < ((config->num_irqs >> 5) + 1); i++) {
```
add a macro for this

carlocaione commented on include/zephyr/drivers/interrupt_controller/riscv_plic.h:
```diff
@@ -12,6 +12,15 @@
+ *
+ * @param dev PLIC device instance.
+ */
+typedef void (*riscv_plic_irq_config_func_t)(const struct device *dev);
```
why this is in a generic public interface? Is it going to be used by users?

ycsin replied:
not at the moment, and not sure if it's ever gonna be used by users, just following the pattern of [uart.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/uart.h#L153)

ycsin replied:
had another look, seem rather unnecessary, removed

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -21,24 +21,39 @@
+/**
+ * @brief Determine the PLIC device from the IRQ
+ *
+ * FIXME: This function is currently hardcoded to return the first instance.
```
right, but this is also making the whole PR basically useless because you cannot match the IRQ number to the actual instance. I don't think it makes sense to merge this without having a fix to this.

ycsin replied:
Yes you are right, this is basically a refactor/cleanup, I was hoping that we can focus on getting the `get_plic_dev_from_irq` right in the other PRs

carlocaione replied:
I see but the problem is that we are risking to merge something half-baked without the certainty that the other PRs will be merged and this fixed. Is it too much work to push a fully working PR with all the needed pieces to have this working?

cfriedt commented:
@ycsin - can you please submit the telink fix in a separate PR? I don't think there should be any issues getting that merged.

ycsin commented:
> telink fix in a separate PR

#62542

I also made another one for the `trig` calculation: #62541

cfriedt commented on drivers/interrupt_controller/Kconfig.plic:
```diff
@@ -10,9 +10,3 @@ config PLIC
 	  Platform Level Interrupt Controller provides support
 	  for external interrupt lines defined by the RISC-V SoC.
-
-config PLIC_SUPPORTS_EDGE_IRQ
```
It isn't clear to me from [the spec](https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc) whether support for level and edge triggers is optional. So from that perspective, it makes sense to remove this option.

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -73,8 +99,10 @@ static int riscv_plic_is_edge_irq(uint32_t irq)
-	volatile uint32_t *en = (volatile uint32_t *)PLIC_IRQ_EN;
 
 	key = irq_lock();
 	en += (irq >> 5);
```
would be good to factor-out the 5 - it seems a little magic-number-y

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -113,7 +143,9 @@ void riscv_plic_irq_disable(uint32_t irq)
+	const struct plic_config *config = dev->config;
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
 
 	en += (irq >> 5);
```
magic number

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -94,8 +122,10 @@ void riscv_plic_irq_enable(uint32_t irq)
-	volatile uint32_t *en = (volatile uint32_t *)PLIC_IRQ_EN;
 
 	key = irq_lock();
 	en += (irq >> 5);
```
magic number

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -113,7 +143,9 @@ void riscv_plic_irq_disable(uint32_t irq)
+	volatile uint32_t *en = (volatile uint32_t *) config->irq_en;
 
 	en += (irq >> 5);
 	return !!(*en & (1 << (irq & 31)));
```
this seems fairly magic-number-y

cfriedt commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -208,43 +241,67 @@ static void plic_irq_handler(const void *arg)
+
+#define PLIC_INTC_INIT_TRIG_0(n) 0
+#define PLIC_INTC_INIT_TRIG_1(n) DT_INST_REG_ADDR_BY_NAME(n, trig)
+#define PLIC_INTC_INIT_TRIG(n) \
```
Assuming that trigger type registers are mandatory, can't we simply derive the exact instance-specific address of `trig` from the value of `num_irqs`? Probably don't even need a DT binding or Kconfig value in that case.

cfriedt reviewed (commented):
Just some commentary. I think this can be simplified a bit, but I would suggest keeping it in draft until other changes have been accepted.

carlocaione commented on drivers/interrupt_controller/intc_plic.c:
```diff
@@ -21,24 +21,47 @@
+{
+	const struct plic_config *config = dev->config;
+
+	return (config->num_irqs >> 5) + 1;
```
yup, magic here as well :)

ycsin commented:
This refactoring is causing `qemu_riscv32_xip` to fail in `tests/kernel/xip`
CI passed because the test wasnt triggered, but was triggered in #62481
I'm guessing that the removal of hardcoded stuff from the interrupt controller probably doesnt play well with XIP

ycsin committed:
tests: build_all: plic: relocate and cleanup

The test is for intc_plic only, so move them into the
intc_plic folder.

Also cleaned up the testcase.yaml a bit and updated the filter
to cover more platforms while I'm at it

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: define trig reg offset in driver, enable with DT

Define the trigger type register offset directly in the driver
and enable it using DT prop instead of Kconfig.

Updated build_all testcase.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: define all register offset in the driver

Define all the register offset directly in the driver according
to the spec. Updated devicetrees that has PLIC accordingly.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
drivers: intc: plic: initial refactor for multi-instance support

Use a config struct to store per-instance device config during
init and connect the IRQ based on the devicetree instead of
hardcoded value and instance number.

The `get_plic_dev_from_irq` is still a placeholder for now and
always return the first instance.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
